components/inputs/input-number.js

```diff
@@ -239,21 +264,49 @@ class InputNumber extends SkeletonMixin(FormElementMixin(LocalizeCoreElement(Lit
 	}
 
 	async _handleChange(e) {
+
 		const value = e.target.value;
 		this._formattedValue = value;
 		await this.updateComplete;
-		const oldValue = this.value;
+
+		let dispatchEvent = false;
 		if (this.trailingZeroes) {
-			this._numDecimalDigits = countDecimalDigits(value);
+			const oldValueTrailingZeroes = this.valueTrailingZeroes;
```

Comment:
This code runs when the user-entered values changes:
- Parse it into a `Number`
- Count how many decimals there were in the value they typed in (in their locale)
- Use the `Intl.NumberFormat` utility to convert that back into a string, but preserve the number of decimals
- Optionally trigger a `change` event